base/shell.jl

```diff
@@ -312,3 +312,51 @@ julia> println(shell_escaped_winsomely("A B\\", "C"))
 """
 shell_escape_winsomely(args::AbstractString...) =
     sprint(print_shell_escaped_winsomely, args..., sizehint=(sum(length, args)) + 3*length(args))
+
+function print_shell_escaped_CMDly(io::IO, arg::AbstractString)
+    any(c -> c in ('\r', '\n'), arg) && throw(ArgumentError("Encountered unsupported character by CMD."))
+    # include " so to avoid toggling behavior of ^
+    arg = replace(arg, r"[()!^\"<>&|]" => s"^\0")
+    arg = replace(arg, r"%" => "%%")
+    print(io, arg)
+end
+
+"""
+     shell_escape_CMDly(arg::AbstractString)  -> String
+
+The unexported `shell_escape_CMDly` function takes a string and escapes any special characters
+in such a way that it is safe to pass it as an argument to some `CMD.exe`. This may be useful
+in concert with the `windows_verbatim` flag to [`Cmd`](@ref) when constructing process
+pipelines.
+
+See also [`shell_escape_PWSHly`](@ref).
+
+# Example
+```jldoctest
+julia> println(shell_escape_CMDly("\"A B\\\" & C"))
+^"A B\\^" ^& C
+
+!important
+    Due to a peculiar behavior of the CMD, each command after a literal `|` character
+    (indicating a command pipeline) must have `shell_escape_CMDly` applied twice. For example:
+    ```
+    to_print = "All for 1 & 1 for all!"
+    run(Cmd(Cmd(["cmd /c \"break | echo \$(shell_escape_CMDly(shell_escape_CMDly(to_print)))"]), windows_verbatim=true))
+    ```
+"""
+shell_escape_CMDly(arg::AbstractString) = sprint(print_shell_escaped_CMDly, arg)
+
+function print_shell_escaped_PWSHly(io::IO, arg::AbstractString)
+    arg = replace(arg, r"[\$#;]" => s"`\0")
+    arg = replace(arg, r"[\"|\\\\(?=\\\\*(\"|$))]" => s"\\\0")
```

Comment:
https://stackoverflow.com/questions/44981177/run-powershell-custom-function-in-new-window/44986341#44986341

Reply:
That left me even more confused. It ends with but it’s not this simple, and the rules have changed over time. 

One thing I’m fairly confident of is that this pwsh function needs us to add \r and \n to the winsomely whitespace list to pass through correctly here (no effect if it doesn’t go to pwsh, so might as well always do it there).

Reply:
Hmm, yeah I'm confused now as well too, I don't think we need it.

> One thing I’m fairly confident of is that this pwsh function needs us to add \r and \n to the winsomely whitespace list to pass through correctly here (no effect if it doesn’t go to pwsh, so might as well always do it there).

Why's that?

Reply:
Initially, I thought we needed powershell to work for this example
```
# works with the added regex
shell> gci C:\\Program\ Files
```
why `\ `? well hitting tab-complete with `shell> gci C:\\Program` spits out `C:\\Program\ Files`, so I figured this should work. The regex fixes this.

BUT the escape for white-spaces is ``` `    ``` , so this doesn't seem necessary. 

Is it at all a odd we have to julia escape backslashes in `shell>` mode or is that normal.

Reply:
FYI updated my comments.

Reply:
@vtjnash  Please see

> The string needs escaping, however, in order to be passed to the new Powershell process unmodified.
> " instances embedded in the string are stripped, unless they are either represented as \" or tripled (""").
> (\" rather than the usual `" is needed to escape double quotes in this case, because PowerShell expects \" when passing a command to the powershell.exe executable.)


I feel confident that we now have a correct escaping solution to powershell and also for CMD

Reply:
I wasn't sure what your source was for `gci`, so I wrote a simple one to be able to inspect what reaches the target application. Save as `gci.c`, compile with `cc gci.c`, then run as `a.exe`.
```
#include <windows.h>
#include <stdio.h>
int main (int argc, char **argv) {
  printf("%s\n", GetCommandLineA());
  for (int i = 0; i < argc; i++) { printf("%d: %s\n", i, argv[i]); }
}
```